model/histogram/float_histogram.go

```diff
@@ -192,6 +192,8 @@ func (h *FloatHistogram) Scale(factor float64) *FloatHistogram {
 //
 // This method returns a pointer to the receiving histogram for convenience.
 func (h *FloatHistogram) Add(other *FloatHistogram) *FloatHistogram {
+	// TODO(trevorwhitney): If other.CounterResetHint != h.CounterResetHint then
+	// we should return some warning.
```

Comment:
Thinking about this again, I think there are legitimate cases, which then require `h.CounterResetHint` to be adjusted. How about adding a switch with approximately the following shape:
- In case `other.CounterResetHint == h.CounterResetHint`: That's the benign case, all good, nothing to do.
- In case one `CounterResetHint` is `GaugeType` and the other is something else: This might be seen as adding an increment to a gauge histogram, which sounds like a legitimate use case. So let's allow it, but the outcome is then a gauge histogram, so `h.CounterResetHint` has to be set to `GaugeType` (in case it isn't already).
- In case one `CounterResetHint` is `UnknownCounterReset` and the other is any other counter-type hint: That might still be legitimate, assuming the caller knows what they are doing. But `h.CounterResetHint` has to be set to `UnknownCounterReset` (in case it isn't already).
- Which leaves us with the remaining case, i.e. `CounterReset` on one and `NotCounterReset` on the other. That cannot be right, and in this case we should return a warning once we can (i.e. here we can leave the TODO above). We should also set `h.CounterResetHint` to `UnknownCounterReset` as the most conservative setting.